contracts/RelayHub.sol

```diff
@@ -232,50 +203,41 @@ contract RelayHub is IRelayHub {
-        bytes4 functionSelector = LibBytesV06.readBytes4(relayRequest.encodedFunction, 0);
-        // Initial soundness checks - the relay must make sure these pass, or it will pay for a reverted transaction.
-        // The worker must be controlled by a manager with a locked stake
+        RelayCallData memory vars;
+        vars.initialGas = gasleft();
+        vars.functionSelector = LibBytesV06.readBytes4(relayRequest.encodedFunction, 0);
+
         require(workerToManager[msg.sender] != address(0), "Unknown relay worker");
         require(
             stakeManager.isRelayManagerStaked(workerToManager[msg.sender], MINIMUM_STAKE, MINIMUM_UNSTAKE_DELAY),
             "relay manager not staked"
         );
-        // A relay may use a higher gas price than the one requested by the signer (to e.g. get the transaction in a
-        // block faster), but it must not be lower. The recipient will be charged for the requested gas price, not the
-        // one used in the transaction.
-        require(relayRequest.gasData.gasPrice <= tx.gasprice, "Invalid gas price");
-        string memory recipientContext;
-        GSNTypes.GasLimits memory gasLimits;
-        {
-            uint256 maxPossibleGas;
-            (maxPossibleGas, gasLimits) = getAndValidateGasLimits(initialGas, relayRequest.gasData, relayRequest.relayData.paymaster);
 
-            // We now verify the legitimacy of the transaction (it must be signed by the sender, and not be replayed),
-            // and that the paymaster will agree to be charged for it.
-            bool success;
-            (success, recipientContext) =
-            // TODO: this new RelayRequest is needed because solc doesn't implement calldata to memory conversion yet
+        require(relayRequest.gasData.gasPrice <= tx.gasprice, "Invalid gas price");
+        // We now verify that the paymaster will agree to be charged for the transaction.
+        bool success;
+        (success, vars.recipientContext, vars.gasLimits) =
             canRelay(
                 GSNTypes.RelayRequest(
                     relayRequest.target,
                     relayRequest.encodedFunction,
                     relayRequest.gasData,
                     relayRequest.relayData),
-                maxPossibleGas, gasLimits.acceptRelayedCallGasLimit, signature, approvalData);
+                    vars.initialGas, signature, approvalData);
 
-            if (!success) {
-                emit CanRelayFailed(
-                    workerToManager[msg.sender],
-                    msg.sender,
-                    relayRequest.relayData.senderAddress,
-                    relayRequest.target,
-                    relayRequest.relayData.paymaster,
-                    functionSelector,
-                    recipientContext);
-                return;
-            }
+        if (!success) {
+            revertReason = GsnUtils.getError(vars.recipientContext);
+            emit UnpaidPaymasterRejection(
```

Comment:
success is "TransactionRelayed".
so failure should be "TxFailed", or TransactionRejectedByPaymaster"

Reply:
We agreed on this name some time ago in the chat.

Reply:
The reason for this event is matching TransactionRelayed and TransactionFailed to tell if a there's a possibility of attack, by someone. That's why its important to make these events close to each other (both in names, and with indexed params)

Reply:
TransactionRelayed is a wrong name in such case, as it indicates success but this is not the case. Will not rename now. Can be streamlined in a separate commit.

Reply:
(also, I don't remember seeing this name agreed upon in any chat)
@shahafn ?